### DISCUSSION TITLE
Conda Package

### DISCUSSION THREAD

**Discussion by filmor** [pythonnet/pythonnet#1864] [General]
@m-rossi I just noticed that both the `pythonnet` and `clr_loader` conda packages are built separately for all operating systems. Is this intentional? Both of them should be buildable on all operating systems, and the build output should be usable on all systems as well (as well as with all supported Python versions). Additional work was required for older versions, but you should now be able to just use (or build) the wheel with only a somewhat recent version of the `dotnet` SDK installed. The result should be `build: noarch: python` as well.

**Comment by m-rossi** [comment]
@filmor I thought about that too when I packaged [clr-loader](https://github.com/conda-forge/clr_loader-feedstock). However [the conda-forge docs](https://conda-forge.org/docs/) are quite clear if a package applies for a `noarch: python`-package:

> In order to qualify as a noarch python package, all of the following criteria must be fulfilled:
> * No compiled extensions

Source: https://conda-forge.org/docs/maintainer/knowledge_base.html#noarch-python

A compiled extension is built both for [clr-loader/setup.py](https://github.com/pythonnet/clr-loader/blob/master/setup.py) and [pythonnet/setup.py](https://github.com/pythonnet/pythonnet/blob/master/setup.py), so I think this applies here.

**Reply by filmor** [reply]
No, it's not a compiled extension in the Pytjon sense. The .NET DLLs that are created here are platform-independent, I just (ab)used the extension mechanism for compiling. The DLLs will work on all systems unchanged (and for clr-loader in particular, the DLL would only be used on Windows, regardless, the rest of the package is pure python.

**Reply by m-rossi** [reply]
I built both packages locally with `noarch: python` in a Linux Docker container and the `conda-build` process seems to be fine. I will test our Windows based company application tomorrow and report back tomorrow. If my tests are successful I will create the PRs in the feedstock-repositories to change the packages to `noarch`.

**Reply by m-rossi** [reply]
Part success. I tested our application, it seems to work with noarch packages.
Therefore I opened both PRs:
* https://github.com/conda-forge/pythonnet-feedstock/pull/16: seems to be fine.
* https://github.com/conda-forge/clr_loader-feedstock/pull/3: There are issues, because I had to use selectors to make it work on Linux and when used in PyPy. The conda-forge-linter requests removal of these selectors.

The problem with PyPy could also be a problem for `pythonnet` too as the [migration for PyPy](https://github.com/conda-forge/pythonnet-feedstock/pull/11) could not yet be merged. I will first try to add support for PyPy in the [rc-branch](https://github.com/conda-forge/pythonnet-feedstock/tree/rc) before merging https://github.com/conda-forge/pythonnet-feedstock/pull/16.

This may also be a general thing to consider: When making these packages noarch, they are only tested on Linux.

**Reply by filmor** [reply]
Python.NET does not support Pypy (check this thread again: https://github.com/pythonnet/pythonnet/issues/330). `clr-loader` does, because it only depends on `cffi` and doesn't actually try to map the Python C API in any way. The `run` dependencies seem a bit off:
* You don't actually require *any* particular .NET runtime, this should be up to the user (they should be optional deps), so I'd just drop `mono` from this one.
* The `__glibc` one I don't understand
* I'd guess that if `cffi` is included in `pypy`, the respective conda package should just be empty, right? I'd just remove that selector

**Reply by m-rossi** [reply]
> You don't actually require any particular .NET runtime, this should be up to the user (they should be optional deps), so I'd just drop mono from this one.

Agreed, we can drop Mono here. Optional dependencies are not possible with conda.

> The __glibc one I don't understand

It was required to be added to user a newer centos image. I will try if it is still necessary.

> I'd guess that if cffi is included in pypy, the respective conda package should just be empty, right? I'd just remove that selector

~~It seems there is no `cffi`-package for `PyPy` at the moment. If I remove the selector the conda package will be incompatible to `PyPy`.~~ Update: Seems you can use `cffi`. I tried to create a new environment with `conda create -n test pypy cffi` and seems to be solvable.

**Reply by m-rossi** [reply]
Success: Both PRs are green now and I will merge.

**Reply by m-rossi** [reply]
@filmor With removing `mono` from the run requirements importing `clr` fails now on the Linux Docker container:
```python
import clr
```

```python
Traceback (most recent call last):
  File "/home/conda/feedstock_root/build_artifacts/pythonnet_1657556318715/test_tmp/run_test.py", line 2, in <module>
    import clr
  File "/home/conda/feedstock_root/build_artifacts/pythonnet_1657556318715/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.10/site-packages/clr.py", line 6, in <module>
    load()
  File "/home/conda/feedstock_root/build_artifacts/pythonnet_1657556318715/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.10/site-packages/pythonnet/__init__.py", line 103, in load
    set_runtime_from_env()
  File "/home/conda/feedstock_root/build_artifacts/pythonnet_1657556318715/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.10/site-packages/pythonnet/__init__.py", line 84, in set_runtime_from_env
    runtime = _create_runtime_from_spec(spec)
  File "/home/conda/feedstock_root/build_artifacts/pythonnet_1657556318715/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.10/site-packages/pythonnet/__init__.py", line 59, in _create_runtime_from_spec
    return clr_loader.get_mono(**params)
  File "/home/conda/feedstock_root/build_artifacts/pythonnet_1657556318715/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.10/site-packages/clr_loader/__init__.py", line 21, in get_mono
    libmono = find_libmono(sgen)
  File "/home/conda/feedstock_root/build_artifacts/pythonnet_1657556318715/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python3.10/site-packages/clr_loader/util/find.py", line 67, in find_libmono
    raise RuntimeError("Could not find libmono")
RuntimeError: Could not find libmono
````

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=534377&view=logs&jobId=656edd35-690f-5c53-9ba3-09c10d0bea97&j=656edd35-690f-5c53-9ba3-09c10d0bea97&t=e5c8ab1d-8ff9-5cae-b332-e15ae582ed2d

**Reply by lostmsu** [reply]
Looking at the error, we could make it a bit friendlier.

**Reply by filmor** [reply]
@m-rossi Well, it fails during the test phase, you need to add the requirement there.
@lostmsu I'm open to suggestions (on `clr_loader`), I was also thinking of adding a nicer `repr` for the loaded runtimes, s.t. one could do a `from pythonnet import get_info(); print(get_info())` to see what is currently loaded from where.

**Reply by m-rossi** [reply]
> @m-rossi Well, it fails during the test phase, you need to add the requirement there.

I think that's no real solution. I expect to install a package with
```shell
conda install pythonnet
```
and do the most basic import
```python
import clr
```
without an error (especially in the conda world). Additional runtimes may be installed separately, but the default runtime should work out of the box.

**Reply by lostmsu** [reply]
@m-rossi can you take a look at [PyTorch](https://pytorch.org/) and see what they do here? They have a CPU-only runtime and a CUDA-based runtime. They have different `conda install` commands with `cudatoolkit` explicitly mentioned for the CUDA one and a probably empty `cpuonly` package for the CPU-only one.

```bash
conda install pytorch cudatoolkit=11.3 -c pytorch
```
```bash
conda install pytorch cpuonly -c pytorch
```

**Reply by m-rossi** [reply]
I think the approach the Torch people use is more complicated than we need here. 

By default `pythonnet` needs `mono` on non-Windows systems. Therefore we should define it as a _requirement_. To do this we need to use a selector. When using selectors we cannot use `noarch: python`.

My recommendation: we should revert the `noarch`-approach and built `pythonnet` for each platform, although it needs more CI-time. I think it's _acceptable_ especially to clearly define the requirements. I will do this in https://github.com/conda-forge/pythonnet-feedstock/pull/17

**Reply by filmor** [reply]
It doesn't need mono. For people that want to use their shiny new .NET 6 library, that's a huge waste. Additionally, IMO it's also a huge waste to build a separate package just because the runtime dependencies differ (are you certain that this is required by conda-forge?). Again, the DLLs built here are /not/ loadable dynamically linked libraries, that's why they are included as package data in the (platform independent) wheels.

**Reply by lostmsu** [reply]
@m-rossi there's no difference that I see between `torch`, `cudatoolkit`, and `cpuonly` packages and `pythonnet`, `netcore`, and `mono` packages correspondingly.

**Reply by m-rossi** [reply]
> It doesn't need mono. For people that want to use their shiny new .NET 6 library, that's a huge waste. Additionally, IMO it's also a huge waste to build a separate package just because the runtime dependencies differ (are you certain that this is required by conda-forge?).

That's the case. Using selectors it is required to build separate packages. But I have another idea in mind using jinja. Let me try a different approach.

> @m-rossi there's no difference that I see between torch, cudatoolkit, and cpuonly packages and pythonnet, netcore, and mono packages correspondingly.

That's true. There are similarities. As said above digging through [their conda-recipes](https://github.com/pytorch/builder/tree/main/conda) I got another idea.

**Reply by m-rossi** [reply]
Quick update: I just removed all runtimes from the run-requirements of `clr-loader` and `pythonnet`. So a we are getting part of the solution the PyTorch people are using. So `pythonnet` including a runtime now has to be installed like this:
```shell
conda install pythonnet mono
conda install pythonnet dotnet-runtime
conda install pythonnet dotnet-runtime=5.0
```
that I am able to test the rc3 in our application and the runtime is also not relevant to me.

In addition I will research my other idea, but I wanted to get out a version that is working right now.

**Reply by filmor** [reply]
To make the second case easier, I'll see if I can adjust clr-loader to automatically use .NET Core if available (e.g. by creating an ad hoc runtineconfig) and falling back to that if Mono is not available.

**Reply by m-rossi** [reply]
I interpreted your statements to mean that you do not want any runtime defined as a requirement. See
* https://github.com/conda-forge/clr_loader-feedstock/blob/main/recipe/meta.yaml#L27-L29
* https://github.com/conda-forge/pythonnet-feedstock/blob/rc/recipe/meta.yaml#L25-L27

Should I add `dotnet-runtime` as a requirement in `clr-loader`? If I leave out a version specification both commands and especially the second command remains possible:
```shell
conda install pythonnet dotnet-runtime
conda install pythonnet dotnet-runtime=5.0
```

**Reply by lostmsu** [reply]
@m-rossi his comment was not relevant for you. He's saying that `clr_loader` without any configuration right now will load Mono if present, but **not** .NET Core. And he wants it to load .NET Core by default if it is present but Mono isn't.

E.g. now (rc3) the behavior without any configuration is:
Mono present -> load Mono
Mono not present -> error

And he wants:
Mono present -> load Mono
Mono and .NET Core present -> load Mono
Mono not present, .NET Core present -> load .NET Core
Nothing present -> error
